### ISSUE TITLE
Build system speedup

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#7]
The Rescuezilla v1.0.5 build system automates the manual [Redo Backup and Recovery Sourceforge Wiki](https://sourceforge.net/p/redobackup/wiki/Home/) instructions. This is a great improvement, but both `make` and `make clean && make` take along time. In non-clean builds, the Ubuntu packages are cached, but Rescuezilla artifacts (such as the chroot) are not.

Given a lot can be developed without rebuilding the ISO image, a relatively slow build is not a deal breaker. However, the build system can and should be able to be improved a lot.

Maybe the best path forward is deconstructing the build.sh, and chroot scripts into a Makefile so a dependency tree is managed by GNU Make. It's also worth considering if alternative approaches such as [Linux Live Kit](https://www.linux-live.org/) manage this problem better.

**Comment by oam7575** [comment]
**TLDR**
Rescuezilla ISO build can be speed reduce to about ~40% with minimal effort depending on your build environment.
Requirements :
CPU Cores : As many as you have
RAM : 6Gb + available for RAM DISK

Using and of the following will speed up the build process a fair amount ; how much depends on the system you are starting with.
* make -j # threads
* a RAM DISK or 
* Compiler Cache : https://ccache.dev/

**First build**
real    23m2.625s
user    30m56.780s
sys     7m15.754s

**Fastest Build**
real    9m31.319s
user    11m35.123s
sys     4m20.116s


**Begins**
OK, so without digging too deep a little bit of testing suggests that the current build system ( July 2021 ) can be speed up a fair amount with some minor adjustments.

What is written below is to try and explain my environment, the changes I made, reasons for them and the results.

My build environment is as follows
**Host**
CPU : Ryzen 5 3600 - 6 cores 12 threads
RAM : 16Gb DDR4 3200
Build Disk :  Samsung 860 QVO 2Tb SATA SSD
OS : Windows 10

**Build Env**
vmware player guest
Linux : Ubuntu 20.04
vCPU : 12
vRAM : 8Gb
vDisk : 200Gb - split files on the 860 QVO


**Build Times**
**Build 1** 
For build 1 I followed the build instructions as per this page : https://github.com/rescuezilla/rescuezilla/blob/master/docs/build_instructions/BUILD.ISO.IMAGE.md to build an ISO _with out_ using Docker.

After updating the Ubuntu vm I set up the build environment as for the workflow and then executed 

"time sudo make focal" resulting in the following
real    23m2.625s
user    30m56.780s
sys     7m15.754s

**Build 2**
The current build scripts make several packages from source using a variety of standard ./configure and make xxx commands.
As make will default to a single thread this seemed a fairly easy place to start.

The Makefile was modified with the following code at the top
THREADS = `cat /proc/cpuinfo | grep process | tail -1 | cut -d":" -f2 | cut -d" " -f2`

This will result in a variable being the number of threads : N -1 ; in my case 11
The THREADS variable was then passed to each instance of the make commands by finding and modifying them.

For example the sfdisk command changed from
        cd $(UTIL_LINUX_BUILD_DIR) && make
to
        cd $(UTIL_LINUX_BUILD_DIR) && make -j $(THREADS)

These changes seemed the simplest and most flexible way to pass thread details to make as it is a small minor change, easy to roll back and should scale with any modern system.

With these changes the build environment was cleaned via "sudo make clean" and "time sudo make focal" was run again.
Resulting build time
real    12m49.794s
user    29m39.126s
sys     5m29.744s


**Build 3**
Several parts of the build process appear to be single threaded a little too deep for easy adjustments at this stage, however file system throughput and compression is usually a bottleneck at some point

As such a test was carried out building rescuezilla on a tmpfs ram disk and backing off the zstd compression part of mkisofs a little.

To achieve this a ram disk was first installed  
sudo mount -t tmpfs -o rw,size=4G tmpfs /mnt/ramdisk/

and the rescuezilla build environment was moved to /mnt/ramdisk.
During the build process the ramdisk ran out of space and the build crashed.

The build environment was cleaned via "sudo make clean" and moved back to "physical disk".

The ram disk was removed, recreated using 6Gb of memory
sudo umount /mnt/ramdisk
sudo mount -t tmpfs -o rw,size=6G tmpfs /mnt/ramdisk/

 and the rescuezilla build environment moved back into ram.

two tests were run with different zstd mkisofs compression values of 6 and 12.

The results are
**zstd = 6**
real    9m46.019s
user    11m58.308s
sys     4m50.256s

ISO size = ~893Mb

**zstd = 12**
real    9m59.098s
user    14m4.859s
sys     4m47.856s

ISO size = ~870Mb

Original ISO size = ~810Mb

**Build 4**
What if we used Compiler Cache to speed to the build process a little more.
Install ccache : sudo aptitude install ccache

Modify the Makefile again, this time changing each instance of the make command to add the ccache settings.
        cd $(UTIL_LINUX_BUILD_DIR) && make -j $(THREADS)
to
        cd $(UTIL_LINUX_BUILD_DIR) && make CC='ccache gcc' -j $(THREADS)

Still maintaining zstd = 12 gives the following
**zstd 12 - CCache Run 1**
real    10m1.017s
user    14m4.703s
sys     4m58.845s

yes - CCache is almost always slightly slower first time round.

**zstd 12 - CCache Run 2**
real    9m31.319s
user    11m35.123s
sys     4m20.116s

This time CCache is slightly faster - however we are still using 11 threads, in a ramdisk on a modern machine; and this builds are not particularly large.

**zstd 12 - CCache Run 3**
What if we roll make back to single thread and see what CCache can do.
For this test -j $(THREADS) was removed to switch make back to single thread mode.

real    9m57.372s
user    11m26.502s
sys     4m3.780s

Note the resulting speed here is about the same as using make -j 11 ; as ccache removes a lot of the compile process.

**Comment by oam7575** [comment]
**EDIT**: For some reason I am having trouble with github failing to escape correctly on my backticks
The "cat /proc .... cut -d" " -f2" part needs to be backtick encapsulated so that the Makefile sends the command out to the OS environment and gets a sensible answer.

**Comment by oam7575** [comment]
Building the Hirsute instance takes a bit more space.
Haven't looked to closely at it - probably some packages have grown with newer releases.
The tmpfs RAMDISK needs to be 6.5~7Gb to allow building in ram.

Seems to work OK.

**Comment by shasheene** [comment]
> **EDIT**: For some reason I am having trouble with github failing to escape correctly on my backticks

Enclosing content in triple backticks (eg. \`\`\` content \`\`\`) should work:
```
content with some random ` backticks asdf and more ` backticks
```

**Comment by oam7575** [comment]
I think I have it edited correctly now.
Sure you'd be able to figure out the command anyways.

**Comment by oam7575** [comment]
Made it a little faster.

time sudo make focal
real    6m39.672s
user    20m4.918s
sys     2m32.440s

So that was a complete focal build at zstd compression 19.
Couple of caveats so that result.

* That speed is a "Hot - second run"
* tmpfs RAMDISK was used as previously described
* Win10 Host power management was configured to force high CPU states ( maintain high Ghz speed and not to idle )
* CCacher was fully warmed up from first run
* I am using my local Apt-Cacher-NG proxy.
* Apt-Cacher was fully populated from first run; local Gbit network; no downloading packages from internet.
* Pigz was pushed into the chroot environment early in the build scripts ( just after the dbus install )
* Pigz was configured as gzip alternative in chroot using update-alternatives.
* Pigz helps a fair amount with the update-initramfs
* I _think_ Pigz helps a little with the language locale compile.

 
This time is a 3rd run.
The process is fairly repeatable.
real    6m34.537s
user    19m40.027s
sys     2m25.128s

**Comment by shasheene** [comment]
Great work, @oam7575! That looks like a great speedup!

It sounds like you have modified the `build.sh` file (and possible some others). Can you please upload your updated build.sh file and I'll take a look!

**Comment by oam7575** [comment]
Pull request should be there
I have left a configure_mirrors.sh script separate as I wasn't sure how you might want to handle that and any potential ARM64 or i686 build, but should be easy enough to modify and integrate as you see fit.

**Comment by shasheene** [comment]
Yep, that's for Pull Request #245! Again, great work! :D

**Comment by oam7575** [comment]
This issue can probably be closed now that you have integrated CCACHE and multithreading to the build.

**Comment by shasheene** [comment]
True!
